File: codigo/p2.java
/**
  Programa 2
  
  Multiiplicar duas matrizes quadráticas de dimensões NxN.
 */

import java.util.Random;

public class p2 {

  public static void main (String[] args) {
    
    int n = Integer.parseInt(args[0]);
    
    float[][] m1 = new float[n][n];
    float[][] m2 = new float[n][n];
    Random rng = new Random();
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        m1[i][j] = rng.nextFloat();
        m2[i][j] = rng.nextFloat();
      }
    }

    Matriz a = new Matriz(m1);
    Matriz b = new Matriz(m2);
    a.multiplicaMatriz(b).printMatriz();
    
  }
  
}


class Matriz {
  
  private float[][] valores;
  
  public Matriz(float[][] valores) {
    this.valores = valores;
  }
  
  public Matriz(int m, int n) {
    this.valores = new float[m][n];
  }
  
  public void printMatriz() {
    for (float[] linha : this.valores) {
      for (float e : linha) {
        System.out.print(e + "\t");
      }
      System.out.println();
    }
  }
  
  public Matriz multiplicaMatriz(Matriz b) {
    

    int l1 = this.valores.length;
    int c1 = this.valores[0].length;
    int l2 = b.valores.length;
    int c2 = b.valores[0].length;

    Matriz m = new Matriz(l1, c2);
  
    // if c1 == l2
    
    for (int i = 0; i < l1; i++) {
      for (int j = 0; j < c2; j++) {
        m.valores[i][j] = 0;

        for (int a = 0; a < c1; a++) {
          m.valores[i][j] += (this.valores[i][a] * b.valores[a][j]);
        }
      }
    }
    
    return m;
  }
  
}

File: codigo/p2.pl
/*
  Programa 2
  
  Multiiplicar duas matrizes quadráticas de dimensões NxN.
 */

% primeira coluna de uma matriz
primeira_coluna([], [], []).
primeira_coluna([[]|_], [], []).
primeira_coluna([[ElemCol|OutrosElem]|OutrasLinhas], [ElemCol|RestoPrimColuna], [OutrosElem|RestoMatriz]) :-
    primeira_coluna(OutrasLinhas, RestoPrimColuna, RestoMatriz).
    
% transposta de uma matriz
transposta([[]|_], []) :- !.
transposta([[A|As]|ARs], [Col|BRs]) :-
    primeira_coluna([[A|As]|ARs], Col, [As|NARs]),
    transposta([As|NARs], BRs).

% produto interno de dois vetores
produto_interno([], [], 0).
produto_interno([A|As], [B|Bs], Soma) :-
    produto_interno(As, Bs, Subtotal),
    Soma is Subtotal + (A*B).


/** <examples>
?- maplist(produto_interno, [[1,2],[3,4]], [[1,2],[3,4]], X).
*/

% multiplicação de duas matrizes
multimatriz([], _, []).
multimatriz([A|As], B, [M|Ms]) :-
    transposta(B, T),
    multimatrizl(A, T, M),
    multimatriz(As, B, Ms).

% predicado auxiliar para multimatriz vetor de produtos internos de um vetor
% com todos as linhas de uma matriz
multimatrizl(_,[],[]) :- !.
multimatrizl(L, [B|Bs], [M|Ms]) :- 
    produto_interno(L, B, K),
    M is K,
    multimatrizl(L, Bs, Ms).


/** <examples>
?- multimatriz([[1,2],[3,4]], [[1,2],[3,4]], X).
*/

% geração de um vetor com números aleatórios
vetor_aleatorio(0, []).
vetor_aleatorio(N, [V|Vs]) :-
    N > 0,
    N1 is N-1,    
    random(V),
    vetor_aleatorio(N1, Vs).

% geração de uma matriz com números aleatórios
matriz_aleatoria(0, _, []).
matriz_aleatoria(L, C, [[M|Ms]|MRs]) :-
    L > 0, C > 0,
    L1 is L-1,
    vetor_aleatorio(C, [M|Ms]),
    matriz_aleatoria(L1, C, MRs).

% geração de uma matriz quadrada com números aleatórios
matriz_quad_aleatoria(N, M) :- matriz_aleatoria(N, N, M).


/** <examples>
?- matriz_quad_aleatoria(100, A), multimatriz(A,A, X)
*/

% predicado para escrever uma matriz na tela
print_matriz([]).
print_matriz([H|T]) :- write(H), nl, print_matriz(T).


main(Argv) :-
    concat_atom(Argv, SingleArg),
    term_to_atom(Term, SingleArg),
    matriz_quad_aleatoria(Term, M1),
    matriz_quad_aleatoria(Term, M2),
    multimatriz(M1, M2, MR),
    print_matriz(MR), halt.
